src/Weather/Skysight/Request.cpp

```diff
@@ -0,0 +1,277 @@
+  return ct;
+}
+
+void
+SkysightAsyncRequest::SetCredentials(const TCHAR *_key,
+				     const TCHAR *_username,
+				     const TCHAR *_password)
+{
+  std::lock_guard<Mutex> lock(mutex);
+  SkysightRequest::SetCredentials(_key, _username, _password);
+}
+
+void
+SkysightAsyncRequest::TriggerNullCallback(tstring &&ErrText)
+{
+  if(args.cb)
+    args.cb(ErrText.c_str(), false, args.layer.c_str(), args.from);
+}
+
+bool
+SkysightRequest::Process()
+{
+  return RequestToFile();
+}
+
+tstring
+SkysightAsyncRequest::GetMessage()
+{
+  std::lock_guard<Mutex> lock(mutex);
+  tstring msg = tstring (_T("Downloading ")) + args.layer;
+  return msg;
+}
+
+bool
+SkysightRequest::ProcessToString(tstring &response)
+{
+  return RequestToBuffer(response);
+}
+
+void
+SkysightAsyncRequest::Process()
+{
+  std::lock_guard<Mutex> lock(mutex);
+  if (IsBusy()) return;
+  Trigger();
+}
+
+void
+SkysightAsyncRequest::Tick() noexcept
+{
+  status = Status::Busy;
+
+  mutex.unlock();
+
+  bool result;
+  tstring resultStr;
+
+  if (args.to_file) {
+    result = RequestToFile();
+    resultStr = args.path.c_str();
+  } else {
+    result = RequestToBuffer(resultStr);
+  }
+
+  if (result) {
+    SkysightAPI::ParseResponse(resultStr.c_str(), result, args);
+  } else {
+    SkysightAPI::ParseResponse(_T("Could not fetch data from Skysight server."),
+			       result, args);
+  }
+
+  mutex.lock();
+  status = (!result) ? Status::Error : Status::Complete;
+}
+
+void
+SkysightAsyncRequest::Done()
+{
+  StandbyThread::LockStop();
+}
+
+bool
+SkysightRequest::RequestToFile()
+{
+  LogFormat("Connecting to %s for %s with key %s", args.url.c_str(), args.path.c_str(), key.c_str());
+
+  Path final_path = Path(args.path.c_str());
+  AllocatedPath temp_path = final_path.WithExtension(".dltemp");
+
+  if (!File::Delete(temp_path) && File::ExistsAny(temp_path))
+    return  false;
+
+  FILE *file = _tfopen(temp_path.c_str(), _T("wb"));
+  if (file == nullptr)
+    return false;
+
+  bool success = true;
+  Net::Session session;
+  FileHandler handler(file);
+  Net::Request request(session, handler, args.url.c_str());
+  request.AddHeader("X-API-Key", key.c_str());
+
+  tstring pBody;
+  if (username.length() && password.length()) {
+    request.AddHeader("Content-Type", "application/json");
+    NarrowString<1024> creds;
+    creds.Format("{\"username\":\"%s\",\"password\":\"%s\"}", username.c_str(), password.c_str());
+    pBody = creds.c_str();
+    request.SetRequestBody(pBody.c_str(), pBody.length());
+    request.SetNoFailOnError();
+  }
+
+  request.SetVerifyPeer(false);
+
+  try {
+    request.Send(10000);
+  } catch (const std::exception &exc) {
+    success = false;
+  }
+
+  success &= fclose(file) == 0;
+
+  if (!success) File::Delete(temp_path);
+
+  file = NULL;
+
+  if (success) {
+    if (!File::Delete(final_path) && File::ExistsAny(final_path)) {
+      File::Delete(temp_path);
+      return false;
+    }
+    std::rename(temp_path.c_str(), args.path.c_str());
```

Comment:
This will fail to compile

Reply:
I build on Android, gcc 7.4.0, gcc 8.3.0. No error. Could you explain more?

Reply:
Build Windows.